sw/device/silicon_creator/lib/epmp_unittest.cc

```diff
@@ -0,0 +1,346 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/epmp.h"
+
+#include <vector>
+
+#include "gtest/gtest.h"
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/csr.h"
+#include "sw/device/lib/base/testing/mock_csr.h"
+
+namespace epmp_unittest {
+using ::mock_csr::CsrTest;
+using ::testing::Test;
+
+/**
+ * PMP control registers.
+ */
+struct PmpCsrs {
+  std::vector<uint8_t> pmpcfg = std::vector<uint8_t>(kEpmpNumRegions, 0);
+  std::vector<uint32_t> pmpaddr = std::vector<uint32_t>(kEpmpNumRegions, 0);
+  uint64_t mseccfg = 0;
+
+  epmp_state_t State() const {
+    epmp_state_t state;
+    for (int i = 0; i < kEpmpNumRegions; ++i) {
+      bitfield_field32_t field = {.mask = 0xff,
+                                  .index = (static_cast<uint32_t>(i) % 4) * 8};
+      state.pmpcfg[i / 4] = bitfield_field32_write(
+          state.pmpcfg[i / 4], field, static_cast<uint32_t>(pmpcfg.at(i)));
+    }
+    for (int i = 0; i < kEpmpNumRegions; ++i) {
+      state.pmpaddr[i] = pmpaddr.at(i);
+    }
+    state.mseccfg = static_cast<uint32_t>(mseccfg);
+    return state;
+  }
+};
+
+/**
+ * Map from PMP config register index to CSR_REG_PMPCFGN value.
+ */
+const std::vector<uint32_t> kPmpcfgMap = {
+    CSR_REG_PMPCFG0,
+    CSR_REG_PMPCFG1,
+    CSR_REG_PMPCFG2,
+    CSR_REG_PMPCFG3,
+};
+
+/**
+ * Map from PMP address register index to CSR_REG_PMPADDRN value.
+ */
+const std::vector<uint32_t> kPmpaddrMap = {
+    CSR_REG_PMPADDR0,  CSR_REG_PMPADDR1,  CSR_REG_PMPADDR2,  CSR_REG_PMPADDR3,
+    CSR_REG_PMPADDR4,  CSR_REG_PMPADDR5,  CSR_REG_PMPADDR6,  CSR_REG_PMPADDR7,
+    CSR_REG_PMPADDR8,  CSR_REG_PMPADDR9,  CSR_REG_PMPADDR10, CSR_REG_PMPADDR11,
+    CSR_REG_PMPADDR12, CSR_REG_PMPADDR13, CSR_REG_PMPADDR14, CSR_REG_PMPADDR15,
+};
+
+/**
+ * Helper function to call EXPECT_CSR_READ on all pmpcfg registers
+ * returning the values provided.
+ */
+static void ExpectPmpcfgRead(const PmpCsrs &expect) {
+  // Pack configuration entries into registers.
+  std::vector<uint32_t> cfg(kEpmpNumRegions / 4, 0);
+  for (uint32_t i = 0; i < expect.pmpcfg.size(); ++i) {
+    bitfield_field32_t field = {.mask = 0xff, .index = (i % 4) * 8};
+    cfg[i / 4] = bitfield_field32_write(
+        cfg.at(i / 4), field, static_cast<uint32_t>(expect.pmpcfg.at(i)));
+  }
```

Comment:
Any reason for not using `pmpcfg` of `expect.State()` instead?

Reply:
Added a new intermediate function.